### ISSUE TITLE
[Training] Proposal: Implement back propagation algorithm

### ISSUE THREAD

**Issue by pauldog** [microsoft/onnxruntime#16809]
### Describe the issue

The back propagation algorithm could be implemented as an open source github.

This would mean we could load an onnx model and train it instead of having to train it in pytorch or tensorflow.

The algorithm itself is quite simple, but it involves a lot of work to implement the backwards function for each operator.

What are your thoughts on the community creating this addition?

### To reproduce

n/a

### ONNX Runtime Installation

Built from Source

### ONNX Runtime Version or Commit ID

15

### PyTorch Version

1

### Execution Provider

Default CPU

**Comment by baijumeswani** [comment]
ONNX Runtime supports training models already. :)

Please see our training solutions:
1. [Large model training](https://onnxruntime.ai/docs/get-started/training-pytorch.html)
2. [On-Device Training](https://onnxruntime.ai/docs/get-started/training-on-device.html).

As far as building the gradient graph is concerned, it is already a part of the ONNX Runtime repository available [here](https://github.com/microsoft/onnxruntime/blob/210d29b40e3eb23b47751f3b4809616a9d825c33/orttraining/orttraining/core/framework/gradient_graph_builder.h#L83) and exposed in python through our [this utility](https://onnxruntime.ai/docs/api/python/on_device_training/training_artifacts.html).

List of available gradient kernels can be found here: https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/core/graph/gradient_builder.h

**Comment by pauldog** [comment]
> ONNX Runtime supports training models already. :)
> 
> Please see our training solutions:
> 
> 1. [Large model training](https://onnxruntime.ai/docs/get-started/training-pytorch.html)
> 2. [On-Device Training](https://onnxruntime.ai/docs/get-started/training-on-device.html).
> 
> As far as building the gradient graph is concerned, it is already a part of the ONNX Runtime repository available [here](https://github.com/microsoft/onnxruntime/blob/210d29b40e3eb23b47751f3b4809616a9d825c33/orttraining/orttraining/core/framework/gradient_graph_builder.h#L83) and exposed in python through our [this utility](https://onnxruntime.ai/docs/api/python/on_device_training/training_artifacts.html).
> 
> List of available gradient kernels can be found here: https://github.com/microsoft/onnxruntime/blob/main/orttraining/orttraining/core/graph/gradient_builder.h

Sorry I should have specified for C#

**Comment by Craigacp** [comment]
You can train models via the Java API, and it should be possible to port that across to the C# API. The Java training session is [here](https://github.com/microsoft/onnxruntime/blob/main/java/src/main/java/ai/onnxruntime/OrtTrainingSession.java).

**Comment by pauldog** [comment]
> You can train models via the Java API, and it should be possible to port that across to the C# API. The Java training session is [here](https://github.com/microsoft/onnxruntime/blob/main/java/src/main/java/ai/onnxruntime/OrtTrainingSession.java).

That looks interesting. I couldn't work out whether it needed a separate "training model" or whether it just takes the ONNX model and applies back propagation to it to create the training model.

**Comment by Craigacp** [comment]
It needs a separate training model generated from an ONNX model by the Python utility that @baijumeswani linked. That utility calls directly into some C code, so if those C methods were added to the training C API then they could be wrapped from C# (or Java) to get a system which takes in an ONNX model and trains it without using Python at all.

**Comment by GeorgeS2019** [comment]
Add ![image](https://github.com/microsoft/onnxruntime/assets/49812372/9d543550-7570-4fe4-b191-bd0db912e7fa)

**Comment by pauldog** [comment]
Any updates in training API in c#?

**Comment by baijumeswani** [comment]
@pauldog we have C# API for training. Please see the API docs here: https://onnxruntime.ai/docs/api/csharp/api/Microsoft.ML.OnnxRuntime.TrainingSession.html

You can find a C# example here: https://github.com/microsoft/onnxruntime-training-examples/tree/master/on_device_training/desktop/csharp

**Comment by pauldog** [comment]
Which nuget packages do I install for C#? I tried installing Microsoft.ML.OnnxRuntime for DirectML and managed and 
Microsoft.ML.OnnxRuntime.Training but it doesn't seem to work. 😟
